DOC-3367: Add new bundling guide overview page. (#3983)
tinymce/tinymce-docs: modules/ROOT/pages/bundling-guide.adoc

= Bundling {productname} - Overview
:navtitle: Bundling overview
:description_short: Overview of bundling TinyMCE with module bundlers
:description: Overview of bundling TinyMCE with module bundlers including required imports, plugins, and resources
:keywords: bundling, webpack, vite, rollup, modules, imports, es6, commonjs

This guide provides an overview of bundling {productname} with module bundlers such as Webpack, Vite, Rollup, or Browserify. For bundler-specific setup instructions, see xref:introduction-to-bundling-tinymce.adoc[Introduction to bundling {productname}].

[NOTE]
====
When using TinyMCE Cloud, dependencies and additional resources are automatically loaded. The information below applies to self-hosted installations using NPM or ZIP packages.
====

== Module syntax support

{productname} supports both ES6 modules and CommonJS modules:

* **ES6 modules**: Use `+import+` statements (shown in all documentation examples)
* **CommonJS modules**: Use `+require()+` statements

Both syntaxes are supported, but this documentation uses ES6 module syntax in all examples.

> == Required imports

Reviewer: Should we be listing the license key manager in here actually? It's only required _if_ using a premium plugin, but. Still a requirement?

Author: I've added it to the required components partial: "When using premium plugins with a commercial license, the `licensekeymanager` plugin is also required."

> include::partial$module-loading/bundling-required-components.adoc[]

Reviewer: Can't find the file, but there's a line that repeats itself:

A custom or premium icon pack to extend or override the default icons for toolbar buttons, including contextual toolbars. This can be a premium or custom icon pack.

Reviewer: We do call it out super well further down, so it's a nitpick

Author: Removed the redundant "This can be a premium or custom icon pack." from the partial.

=== Example: Basic required imports

[source,js]
----
/* Import TinyMCE */
import tinymce from 'tinymce';

/* Default icons are required */
import 'tinymce/icons/default';

/* Required TinyMCE components */
import 'tinymce/themes/silver';
import 'tinymce/models/dom';

/* Import a skin (oxide is the default) */
import 'tinymce/skins/ui/oxide/skin.js';
import 'tinymce/skins/ui/oxide/content.js';

/* Content CSS for editor content styling */

> import 'tinymce/skins/content/default/content.js';

Reviewer: Said in the section above that this isn't required, then put it in an example of _required_ imports?

Reviewer 2: Maybe we could say 'recommended imports' instead?

Author: Good catch. I've renamed the example to "Example: Basic recommended imports" to align with content CSS being optional.

----

For more information on:

* Icons, see xref:bundling-icons.adoc[Bundling icons]
* Themes, see xref:bundling-themes.adoc[Bundling themes]
* Models, see xref:bundling-models.adoc[Bundling models]
* Skins, see xref:bundling-skins.adoc[Bundling skins]
* Content CSS, see xref:bundling-content-css.adoc[Bundling content CSS]

== Importing plugins

Most plugins can be imported directly using their plugin code. The import path depends on whether the plugin is a community or premium plugin:

> === Community plugins

Reviewer: I do wonder if it's worth mentioning somewhere in this page that we've mostly (entirely?) talking about pulling files from NPM? Like, it's implied but also... The premium example in particular won't make sense if you're thinking of using a bundle or something.

Author: Agreed. I've added a sentence to the NOTE: "All examples in this guide use NPM package import paths."

----
import 'tinymce-premium/plugins/<plugincode>';
----

[IMPORTANT]
====
When using premium plugins with a commercial license, you must include the `+licensekeymanager+` plugin:

[source,js]
----
import 'tinymce-premium/plugins/licensekeymanager';
----

For more information, see xref:bundling-plugins.adoc#using-premium-plugins[Using premium plugins].
====

For detailed information on bundling plugins, see xref:bundling-plugins.adoc[Bundling plugins].

== Plugin gotchas

Some plugins require additional resource files (JS, CSS, or language files) to function properly. These must be explicitly imported when bundling:

=== Plugins requiring additional resources

* **Emoticons**: Requires an emoji database file (`+js/emojis.js+`)
* **Help**: Requires a keyboard navigation language file (`+js/i18n/keynav/<lang>.js+`)
* **Autocorrect**: Requires a word list file (`+js/en.js+`)
* **Uploadcare**: Requires a webcomponent file (`+js/uc-video.js+`)

For specific import examples, see xref:bundling-plugins.adoc#plugin-resources[Additional plugin resources].

=== Plugin dependencies

Some plugins depend on other plugins. For example:

* The `+advtable+` plugin requires the `+table+` plugin

> Always import dependent plugins before the plugin that requires them.

Reviewer: We should probably mention where else to find this info, since we've framed this as an example and not a full list. Should advise to check the plugin's page I guess?

Author: I've added: "For a full list of plugin dependencies, see each plugin's documentation page."

== CSS files

Some plugins require additional CSS files. Plugin CSS files are included automatically when importing plugins via their JS file.

When using skins, you must import both:

* The skin CSS file (`+skin.css+` or `+skin.js+`)

> * The content UI CSS file (`+content.css+` or `+content.js+`)

Reviewer 2: ```suggestion
* The skin CSS file (`+skin.css+` or `+skin.js+`)
* The content skin CSS file (`+content.css+` or `+content.js+`)
```
Just to make sure it's not confused with the Content CSS file.

[NOTE]
====

> Importing the `+skin.js+` and `+content.js+` files is an alternative way to load skin styles. For bundlers like Webpack and Browserify that support CSS imports, the CSS examples in xref:bundling-skins.adoc[Bundling skins] remain valid.

Reviewer 2: ```suggestion
Importing the `+skin.js+` and `+content.js+` files is an alternative way to load skin styles besides CSS imports. For bundlers like Webpack and Browserify that require additional setup to support CSS imports, the CSS examples in xref:bundling-skins.adoc[Bundling skins] remain valid.
```

====

== Module syntax support

{productname} supports both ES6 modules and CommonJS modules:

* **ES6 modules**: Use `+import+` statements (shown in all documentation examples)
* **CommonJS modules**: Use `+require()+` statements

Both syntaxes are supported, but this documentation uses ES6 module syntax in all examples.

== Required imports

include::partial$module-loading/bundling-required-components.adoc[]

=== Example: Basic required imports

[source,js]
----
/* Import TinyMCE */
import tinymce from 'tinymce';

/* Default icons are required */
import 'tinymce/icons/default';

/* Required TinyMCE components */
import 'tinymce/themes/silver';
import 'tinymce/models/dom';

/* Import a skin (oxide is the default) */
import 'tinymce/skins/ui/oxide/skin.js';
import 'tinymce/skins/ui/oxide/content.js';

/* Content CSS for editor content styling */
import 'tinymce/skins/content/default/content.js';
----

For more information on:

* Icons, see xref:bundling-icons.adoc[Bundling icons]
* Themes, see xref:bundling-themes.adoc[Bundling themes]
* Models, see xref:bundling-models.adoc[Bundling models]

> * Skins, see xref:bundling-skins.adoc[Bundling skins]

Reviewer 2: The page for [Bundling skin](https://www.tiny.cloud/docs/tinymce/latest/bundling-skins/) mainly references CSS files. @metricjs I think for guides like Webpack and Browserify that already have ways to import CSS files, it may be worth keeping those examples around. We could consider adding a note that importing the `js` file is another way to get the CSS, but it seems safest not to remove those CSS examples when we start tackling those specific pages.

Author: Updated in last commit

* Content CSS, see xref:bundling-content-css.adoc[Bundling content CSS]

== Importing plugins

Most plugins can be imported directly using their plugin code. The import path depends on whether the plugin is a community or premium plugin:

=== Community plugins

[source,js]
----
import 'tinymce/plugins/<plugincode>';
----

=== Premium plugins

[source,js]
----
import 'tinymce-premium/plugins/<plugincode>';
----

[IMPORTANT]
====
When using premium plugins with a commercial license, you must include the `+licensekeymanager+` plugin:

[source,js]
----
import 'tinymce-premium/plugins/licensekeymanager';
----

For more information, see xref:bundling-plugins.adoc#using-premium-plugins[Using premium plugins].
====

For detailed information on bundling plugins, see xref:bundling-plugins.adoc[Bundling plugins].

== Plugin gotchas

> Some plugins require additional resource files (JS, CSS, or language files) to function properly. These must be explicitly imported when bundling:

Reviewer: ```suggestion
Some plugins require additional resource files (JS, CSS, or language files) to function properly. These must be explicitly imported when bundling.
```
?

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

[IMPORTANT]
====
Key points about plugin resources:

* Core plugin resources typically use relative paths without file extensions

> * Premium plugin resources (from `+tinymce-premium+`) MUST include the `.js` or `.css` file extension in the import path

Reviewer: I can't see this noted in Sorita's notes and wasn't aware it was a thing - sanity checking where this comes from?

Reviewer: Oh, was this from her examples for the lang files?

Reviewer 2: I think we may not need to include this. An integrator could follow their own import styles, so long as their bundler is configured properly to resolve the file extensions. The choices in the example code were mainly to ensure it would work for the majority out of the box.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061) removed

> * Plugins lazy-load additional resources at runtime, so these files must be included in the bundle or accessible as static assets

Reviewer: "Some plugins" or "plugins may" - to not cause panic that every plugin pulls in more resources?

Author: Was removed in[this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

====

=== Plugins requiring additional resources

> * **Emoticons**: Requires an emoji database file (`+js/emojis+`)

Reviewer 2: ```suggestion
* **Emoticons**: Requires an emoji database file (`+js/emojis.js+`)
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

> * **Help**: Requires a keyboard navigation language file (`+js/i18n/keynav/<lang>+`)

Reviewer 2: ```suggestion
* **Help**: Requires a keyboard navigation language file (`+js/i18n/keynav/<lang>.js+`)
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

> * **Page Embed**: Requires a CSS file (`+css/empa30.css+`)

Reviewer 2: From my testing, this import is not necessary. Was there an issue reported on page embed that required this import?

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

* **Autocorrect**: Requires a word list file (`+js/en.js+`)
* **Uploadcare**: Requires a webcomponent file (`+js/uc-video.js+`)

For specific import examples, see xref:bundling-plugins.adoc#plugin-resources[Additional plugin resources].

=== Plugin dependencies

Some plugins depend on other plugins. For example:

* The `+advtable+` plugin requires the `+table+` plugin

Always import dependent plugins before the plugin that requires them.

== CSS files

> Plugin CSS files are typically included automatically when importing plugins. However, some plugins may require explicit CSS imports for proper styling.

Reviewer: ```suggestion
Some plugins require additional CSS files. Plugin CSS files are included automatically when importing plugins via their JS file.
```
or something? from what Sorita said, should never need to pull in a CSS file manually

Reviewer 2: Currently, there aren't any plugins that require a CSS import, and if it does, then it would likely require a fix - as in the case for Uploadcare.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

When using skins, you must import both:

* The skin CSS file (`+skin.css+` or `+skin.js+`)
* The content UI CSS file (`+content.css+` or `+content.js+`)

For more information, see xref:bundling-skins.adoc[Bundling skins].

== Language files

Language files are optional and used for localizing the {productname} UI and plugin interfaces.

=== Core UI language files

> 

Reviewer: ```suggestion

For community packages:

```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

[source,js]
----
import 'tinymce/langs/<language>.js';
----

For premium packages:

[source,js]
----
import 'tinymce-premium/langs/<language>.js';
----

=== Plugin language files

> Plugin language files are separate from the main UI language files:

Reviewer 2: I think we should also note that the editor's `language` option must first be set to the desired language for the plugin's language file to take effect.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

[source,js]
----
// Community plugin language file
import 'tinymce/plugins/<plugincode>/langs/<language>.js';

> // Premium plugin language file (note: .js extension required)

Reviewer 2: ```suggestion
// Premium plugin language file
```
We can remove this

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

import 'tinymce-premium/plugins/<plugincode>/langs/<language>.js';
----

[NOTE]
====

> * English is the default language, so `+en.js+` language files don't exist

Reviewer 2: Maybe worth specifying it's US English? https://www.tiny.cloud/docs/tinymce/latest/ui-localization/#language. Perhaps also mentioning language packs for extra clarity, as `help` and `autocorrect` plugins have an `en.js` file.

```suggestion
* US English is the default language, so the language packs do not include `+en.js+` language files
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3983/commits/2100b71771ad9c4253c03fbbae8a5482c4af3061)

> * For premium plugins, the language file path must include the `.js` extension

Reviewer: From the examples here, it looks like the .js is always required, whether premium or community? @soritaheng any thoughts?

Reviewer 2: Most bundlers support resolving file extensions, so they are not necessarily required. For example, in [Vite](https://vite.dev/config/shared-options#resolve-extensions). Since the order of the resolution could be configured, I went with the extensions in the examples for clarity's sake. But let me know if that should be reconsidered.
